### PR TITLE
Use kubectl in EjsonSecretProvisioner

### PR DESCRIPTION
**Pull request by KnVerey**
tl;dr `kubeclient` isn't working for some of our production clusters, so this switches `EjsonSecretProvisioner` over to `kubectl`--which is what the rest of the deploy process uses--for the time being. 

---

When I rolled out https://github.com/Shopify/kubernetes-deploy/pull/81 in shipit, I discovered that kubeclient does not seem to be working with our GCP clusters. It is failing with an SSL error on `client.discover` ([see this deploy](https://shipit.shopify.io/shopify/hackdays-gke-rails5/production/deploys/445001)). This may be a further piece to the work @xldenis already did in https://github.com/Shopify/kubernetes-deploy/pull/88; he is investigating.

The most meaningful change is that we need to write a tempfile to use `kubectl apply` for secret creation/update. Note that although the unit tests required surgery, the integration tests all passed as-is.

**Why not wait for the kubeclient fix?** 
The current consensus is still that we want to use a gem rather than kubectl for everything we can (which doesn't include "apply" operations) in the long run, but I didn't realize that none of the tasks that already use kubeclient had been tested in our production Shipit environment. This feature is high-priority, and switching to kubeclient is far from it. With the problems we've had, we may even want to reconsider this particular gem choice.

cc @Shopify/cloudplatform

### CODE REVIEW

File: lib/kubernetes-deploy/ejson_secret_provisioner.rb [hunk 1]
```diff
@@ -96,31 +99,45 @@ def validate_secret_spec(secret_name, spec)
...
-        KubernetesDeploy.logger.info("Updating secret #{secret_name}")
-        @kubeclient.update_secret(secret)
-      else
-        KubernetesDeploy.logger.info("Creating secret #{secret_name}")
-        @kubeclient.create_secret(secret)
+      secret_yaml = generate_secret_yaml(secret_name, secret_type, data)
+      msg = secret_exists?(secret_name) ? "Updating secret #{secret_name}" : "Creating secret #{secret_name}"
+      KubernetesDeploy.logger.info(msg)
+
+      file = Tempfile.new(secret_name)
```
Comment (KnVerey) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: Alternatively, we could construct a really long command with `--from-literal=k=v --from-literal=k=v ...`

File: lib/kubernetes-deploy/ejson_secret_provisioner.rb [hunk 2]
```diff
@@ -152,17 +169,28 @@ def decrypt_ejson(key_dir)
...
+    def run_kubectl_json(*args)
+      args += ["--output=json"]
+      out, err, st = run_kubectl(*args)
+      raise EjsonSecretError, err unless st.success?
+      result = JSON.parse(out)
+      result.fetch('items', result)
+    end
+
+    def run_kubectl(*args)
+      raise FatalDeploymentError, "Namespace missing for namespaced command" if @namespace.blank?
```
Comment (camilo) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: Why not check this on the constructor?

File: lib/kubernetes-deploy/ejson_secret_provisioner.rb [hunk 3]
```diff
@@ -152,17 +169,28 @@ def decrypt_ejson(key_dir)
...
+      raise KubectlError, "Explicit context is required to run this command" if @context.blank?
```
Comment (camilo) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: Same here

File: lib/kubernetes-deploy/ejson_secret_provisioner.rb [hunk 4]
```diff
@@ -96,31 +99,45 @@ def validate_secret_spec(secret_name, spec)
...
-        @kubeclient.create_secret(secret)
+      secret_yaml = generate_secret_yaml(secret_name, secret_type, data)
+      msg = secret_exists?(secret_name) ? "Updating secret #{secret_name}" : "Creating secret #{secret_name}"
+      KubernetesDeploy.logger.info(msg)
+
+      file = Tempfile.new(secret_name)
+      file.write(secret_yaml)
+      file.close
+
+      out, err, st = run_kubectl("apply", "--filename=#{file.path}")
```
Comment (mkobetic) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: We should make sure the file is deleted too, would `Tempfile.open` work for this?

File: lib/kubernetes-deploy/ejson_secret_provisioner.rb [hunk 5]
```diff
@@ -96,31 +99,45 @@ def validate_secret_spec(secret_name, spec)
...
-      else
-        KubernetesDeploy.logger.info("Creating secret #{secret_name}")
-        @kubeclient.create_secret(secret)
+      secret_yaml = generate_secret_yaml(secret_name, secret_type, data)
+      msg = secret_exists?(secret_name) ? "Updating secret #{secret_name}" : "Creating secret #{secret_name}"
+      KubernetesDeploy.logger.info(msg)
+
+      file = Tempfile.new(secret_name)
+      file.write(secret_yaml)
+      file.close
```
Comment (KnVerey) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: To address both this and @mkobetic comment below, kubectl won't be able to read the file unless we close it. Tempfiles are automatically GC'd when the process exits, if not sooner. `Tempfile.open { |f| ... }` doesn't return the file. I don't think it would be possible for the file to get GC'd when we need it in this case, since we only need it for the duration of this method, which retrains a reference to it. However, I can switch to a regular file and manually unlink it in an `ensure` to be extra super safe.

Comment (mkobetic) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: I would expect that if the file was flushed after write the shell command should be able to read it then, but I haven't tried.

Comment (KnVerey) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: TBH I think this is the simplest solution. The GC should take care of unlinking at the appropriate time, but I'll add `file.unlink if file` in an `ensure` just in case.

File: lib/kubernetes-deploy/ejson_secret_provisioner.rb (hunk 2 above)
Comment (KnVerey) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: Because this is copied from another place where it wouldn't make sense in the constructor. :P

Really, it's a super-paranoid check to make sure we don't run commands without a context/namespace. In reality, it should be impossible to hit in production (several other things would have raised before we get here). But I don't feel strongly about where it goes and can move it.

**Comment by KnVerey** [comment]
Updated

File: lib/kubernetes-deploy/ejson_secret_provisioner.rb [hunk 6]
```diff
@@ -99,17 +102,19 @@ def validate_secret_spec(secret_name, spec)
...
+      secret_yaml = generate_secret_yaml(secret_name, secret_type, data)
       file = Tempfile.new(secret_name)
       file.write(secret_yaml)
       file.close
 
       out, err, st = run_kubectl("apply", "--filename=#{file.path}")
       KubernetesDeploy.logger.debug(out)
       raise EjsonSecretError, err unless st.success?
+    ensure
+      file.unlink if file
```
Comment (mkobetic) [inline lib/kubernetes-deploy/ejson_secret_provisioner.rb]: I wonder if this can actually error out in some circumstances, I'd be fine to rely on Tempfile to take care of this. But I'm fine either way
